File: app/src/main/java/in/completecourse/fragment/mainFragment/NotificationFragment.java
package in.completecourse.fragment.mainFragment;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import in.completecourse.PDFActivity;
import in.completecourse.R;
import in.completecourse.adapter.NotificationAdapter;
import in.completecourse.app.AppConfig;
import in.completecourse.helper.HelperMethods;
import in.completecourse.helper.HttpHandler;
import in.completecourse.model.NotificationModel;


public class NotificationFragment extends Fragment {

    private ArrayList<NotificationModel> itemsList;
    private NotificationAdapter mAdapter;
    @BindView(R.id.progress_bar)
    ProgressBar progressBar;
    @BindView(R.id.recyclerView_notification)
    RecyclerView recyclerView;
    @BindView(R.id.empty_layout)
    RelativeLayout emptyView;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view =  inflater.inflate(R.layout.fragment_notification, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        itemsList = new ArrayList<>();
        RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(view.getContext(), RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(mLayoutManager);
        recyclerView.addItemDecoration(new DividerItemDecoration(view.getContext(), DividerItemDecoration.VERTICAL));

        recyclerView.setNestedScrollingEnabled(false);

        if (HelperMethods.isNetworkAvailable(getActivity())){
            new GetNotifications(NotificationFragment.this).execute();

        }else{
            Toast.makeText(getContext(), "Please check your internet connection.", Toast.LENGTH_SHORT).show();
        }
    }

    private static class GetNotifications extends AsyncTask<Void, Void, Void> {
        NotificationModel model;
        private final WeakReference<NotificationFragment> activityWeakReference;

        GetNotifications(NotificationFragment context){
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected void onPreExecute() {
            NotificationFragment activity = activityWeakReference.get();
            activity.progressBar.setVisibility(View.VISIBLE);
        }

        @Override
        protected Void doInBackground(Void... arg0) {
            NotificationFragment newArrivalFragment = activityWeakReference.get();
            HttpHandler sh = new HttpHandler();
            String url = AppConfig.URL_NOTIFICATION;
            String jsonStr = sh.makeServiceCall(url);
            if (jsonStr != null) {
                try {
                    JSONObject jsonObject = new JSONObject(jsonStr);
                    JSONArray jsonArray = jsonObject.getJSONArray("data");
                    for (int i = 0; i < jsonArray.length(); i++) {
                        model = new NotificationModel();
                        JSONObject c = jsonArray.getJSONObject(i);
                        model.setmHeading(c.getString("notifyheading"));
                        model.setmSubHeading(c.getString("notifydetails"));
                        model.setUrl(c.getString("notifyURL"));
                        model.setSerial((i + 1) + ". ");
                        newArrivalFragment.itemsList.add(model);
                    }
                } catch (final JSONException e) {

                    Toast.makeText(newArrivalFragment.getActivity(), "Json parsing error: " + e.getMessage(), Toast.LENGTH_LONG).show();
                }
            } else {
                Toast.makeText(newArrivalFragment.getActivity(), "Couldn't get data from server.", Toast.LENGTH_SHORT).show();
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);
            final NotificationFragment notificationFragment = activityWeakReference.get();
            notificationFragment.progressBar.setVisibility(View.GONE);
            if (notificationFragment.itemsList.isEmpty()){
                notificationFragment.emptyView.setVisibility(View.VISIBLE);
            }
            notificationFragment.mAdapter = new NotificationAdapter(notificationFragment.getActivity(), notificationFragment.itemsList);
            notificationFragment.recyclerView.setAdapter(notificationFragment.mAdapter);
            notificationFragment.recyclerView.addOnItemTouchListener(new NotificationAdapter.RecyclerTouchListener(notificationFragment.getContext(), position -> {
                String url = notificationFragment.itemsList.get(position).getUrl();
                Intent intent = new Intent(notificationFragment.getActivity(), PDFActivity.class);
                intent.putExtra("url", url);
                notificationFragment.startActivity(intent);
            }));
        }
    }
}

File: app/src/main/java/in/completecourse/fragment/mainFragment/NotificationFragment.kt
package `in`.completecourse.fragment.mainFragment

import `in`.completecourse.PDFActivity
import `in`.completecourse.adapter.NotificationAdapter
import `in`.completecourse.databinding.FragmentNotificationBinding
import `in`.completecourse.helper.HelperMethods
import `in`.completecourse.model.NotificationModel
import `in`.completecourse.utils.APIService
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import retrofit2.Retrofit
import java.util.*

class NotificationFragment : Fragment(), NotificationAdapter.ClickListener {
    private var itemsList: ArrayList<NotificationModel>? = null
    private var mAdapter: NotificationAdapter? = null

    private var _binding: FragmentNotificationBinding? = null
    //This property is only valid between onCreateView and
    //onDestroyView
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentNotificationBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        itemsList = ArrayList<NotificationModel>()
        val mLayoutManager: RecyclerView.LayoutManager = LinearLayoutManager(view.context, RecyclerView.VERTICAL, false)

        binding.recyclerViewNotification.layoutManager = mLayoutManager
        binding.recyclerViewNotification.addItemDecoration(DividerItemDecoration(view.context, DividerItemDecoration.VERTICAL))
        binding.recyclerViewNotification.isNestedScrollingEnabled = false

        if (HelperMethods.isNetworkAvailable(activity)) {
            getNotifications()
        } else {
            Toast.makeText(context, "Please check your internet connection.", Toast.LENGTH_SHORT).show()
        }
    }

    private fun getNotifications(){
        binding.progressBar.visibility = View.VISIBLE
        //Create Retrofit
        val retrofit = Retrofit.Builder().baseUrl("http://completecourse.in/api/").build()

        //Create Service
        val service = retrofit.create(APIService::class.java)

        CoroutineScope(Dispatchers.IO).launch {
            //Do the POST request and get the response
            val response = service.getNotifications()
            if (response.isSuccessful) {
                var model: NotificationModel
                val obj = JSONObject(response.body()?.string() ?: "{}")
                val jsonArray = obj.getJSONArray("data")//JSONArray(obj.getJSONArray("data"))
                //Log.e("jsonArray", jsonArray.toString())
                for (i in 0 until jsonArray.length()) {
                    model = NotificationModel()
                    val c = jsonArray.getJSONObject(i)
                    model.mHeading = c.getString("notifyheading")
                    model.mSubHeading = c.getString("notifydetails")
                    model.url = c.getString("notifyURL")
                    model.serial = (i + 1).toString() + ". "
                    itemsList?.add(model)
                }
            }

            withContext(Dispatchers.Main){
                binding.progressBar.visibility = View.GONE
                if (itemsList!!.isEmpty()) {
                    binding.emptyLayout.visibility = View.VISIBLE
                }
                mAdapter = NotificationAdapter(activity!!, itemsList!!)
                binding.recyclerViewNotification.adapter = mAdapter
                binding.recyclerViewNotification.addOnItemTouchListener(NotificationAdapter.RecyclerTouchListener(context, this@NotificationFragment))
            }
        }
    }

    override fun onClick(position: Int) {
        val url: String? = itemsList!![position].url
        val intent = Intent(activity, PDFActivity::class.java)
        intent.putExtra("url", url)
        startActivity(intent)
    }
}